ecs-cli/modules/clients/aws/servicediscovery/client.go

```diff
@@ -0,0 +1,153 @@
+	for _, namespaceID := range nameMatch {
+		hasVPC, err := checkNamespaceVPC(namespaceID, vpc, region, config)
+		if err != nil {
+			return nil, err
+		}
+		if hasVPC {
+			matchingNamespaces = append(matchingNamespaces, namespaceID)
+		}
+	}
+
+	return matchingNamespaces, nil
+}
+
+// FindPublicNamespace returns the ID of the public namespace with the given name
+func FindPublicNamespace(name string, config *config.CommandConfig) (*string, error) {
+	var namespace *string
+	err := listNamespaces(true, config, func(n *servicediscovery.NamespaceSummary) bool {
+		if aws.StringValue(n.Name) == name {
+			namespace = n.Id
+			return false // we found it, stop the list call
+		}
+		return true
+	})
+
+	return namespace, err
+}
+
+func checkNamespaceVPC(namespaceID *string, vpc string, region string, config *config.CommandConfig) (bool, error) {
+	namespaceInfo, err := getNamespace(namespaceID, config)
+	if err != nil {
+		return false, err
+	}
+	hostedZoneID := namespaceInfo.Namespace.Properties.DnsProperties.HostedZoneId
+	hostedZone, err := getHostedZone(hostedZoneID, config)
+	if err != nil {
+		return false, err
+	}
+	for _, hostedZoneVPC := range hostedZone.VPCs {
+		// The VPC must be in the region that we will be launching the ECS Service
+		if (aws.StringValue(hostedZoneVPC.VPCId) == vpc) && (aws.StringValue(hostedZoneVPC.VPCRegion) == region) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// ServiceDiscovery API Actions that can be mocked in unit tests
+var sdClient *servicediscovery.ServiceDiscovery
+
+func getSDClient(config *config.CommandConfig) *servicediscovery.ServiceDiscovery {
+	if sdClient == nil {
+		sdClient = servicediscovery.New(config.Session)
+		sdClient.Handlers.Build.PushBackNamed(clients.CustomUserAgentHandler())
+	}
+	return sdClient
+}
+
+// Lists namespaces, calling 'fn' on each namespace returned.
+// To stop iterating over namespaces, return false from 'fn'
+var listNamespaces = func(isPublic bool, config *config.CommandConfig, fn func(*servicediscovery.NamespaceSummary) bool) error {
+	client := getSDClient(config)
+
+	typeFilter := servicediscovery.NamespaceTypeDnsPrivate
+	if isPublic {
+		typeFilter = servicediscovery.NamespaceTypeDnsPublic
+	}
+	request := &servicediscovery.ListNamespacesInput{
+		Filters: []*servicediscovery.NamespaceFilter{
+			&servicediscovery.NamespaceFilter{
+				Condition: aws.String(servicediscovery.FilterConditionEq),
+				Name:      aws.String(servicediscovery.NamespaceFilterNameType),
+				Values:    aws.StringSlice([]string{typeFilter}),
+			},
+		},
+	}
+	err := client.ListNamespacesPages(request,
+		func(page *servicediscovery.ListNamespacesOutput, lastPage bool) bool {
+			for _, namespace := range page.Namespaces {
+				if !fn(namespace) {
+					return false
+				}
+			}
+			return !lastPage
+		})
+	return err
+}
+
+var getNamespace = func(id *string, config *config.CommandConfig) (*servicediscovery.GetNamespaceOutput, error) {
+	client := getSDClient(config)
```

Comment:
I still feel inclined to have the clients isolated somewhere more obvious -- I'm not sure what the value is in obscuring these calls; seems like blurring that isolation boundary would potentially make things difficult to debug if there is a failure on the client.

Reply:
Where would you prefer to have the clients isolated/located?

Reply:
I guess my inclination is that anything that is calling SDS would be in an SDS package, for example, so it's very clear what methods are interacting with which outside services. I find it hard to read/understand the code when I have to trace the code this car down to get to the client call, but I'm willing to defer this to a later refactor since none of this is actually wired in yet.